### PR TITLE
feat: Safety Mode for Runtime

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- Add safety mode for Torch-TensorRT runtime
- Add C++ TorchBind bindings and relevant lambda functions to get and set necessary attributes
- Add runtime augmentations to support different modes
- Add Python-runtime device checks
- Add testing for safe mode settings

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Review by peri044 — APPROVED** [review]

File: tests/py/dynamo/runtime/test_safe_mode.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        fx_graph = torch.fx.symbolic_trace(SampleModel())
+
+        # Validate that the results between Torch and Torch-TRT are similar
+        optimized_model = torch_tensorrt.compile(
+            fx_graph,
+            "torch_compile",
+            inputs,
+            min_block_size=1,
+            pass_through_build_failures=True,
+            use_python_runtime=True,
```
[the remark this answers is not in the document]
Comment (gs-olive) [reply tests/py/dynamo/runtime/test_safe_mode.py:42]: Thanks for the catch - now added 2 test cases, one for each runtime

File: core/runtime/runtime.cpp
Comment (gs-olive) [inline core/runtime/runtime.cpp:1]: When loading engine, recommend running in `SAFE_MODE` with a log warning if multiple GPUs are detected

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:1]: Can be serialized to `state_dict`, add warning to `load` when on a multi-GPU system.

File: core/runtime/runtime.cpp [hunk 2]
```diff
@@ -7,6 +7,8 @@ namespace torch_tensorrt {
 namespace core {
 namespace runtime {
 
+bool SAFE_MODE = true;
```
Comment (gs-olive) [inline core/runtime/runtime.cpp:10]: Rename to `MULTI_DEVICE_SAFE_MODE`

File: py/torch_tensorrt/_compile.py [hunk 3]
```diff
@@ -256,6 +258,26 @@ def torch_compile(module: torch.nn.Module, **kwargs: Any) -> Any:
...
+
+
+def enable_safe_inference_mode() -> None:
+    """
+    Enables safe inference mode for Torch-TensorRT
+    """
+    global SAFE_MODE
+    SAFE_MODE = True
+    torch.ops.tensorrt.set_safe_mode(True)
+    logger.info("Enabled safe inference mode")
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:278]: `set_multi_device_safe_mode(bool mode)`

Comment (gs-olive) [reply py/torch_tensorrt/_compile.py:278]: `with set_multi_device_safe_mode(True):` context wrapper

File: core/runtime/execute_engine.cpp
Comment (gs-olive) [inline core/runtime/execute_engine.cpp:1]: Warning @ deserialization time / engine instantiation

File: py/torch_tensorrt/_compile.py [hunk 4]
```diff
@@ -20,6 +20,8 @@
 
 DYNAMO_ENABLED = version.parse(sanitized_torch_version()) >= version.parse("2.1.dev")
 
+SAFE_MODE = torch.ops.tensorrt.get_safe_mode()
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:23]: Need to check for `torch_tensorrt._C` import before setting this. `_PY_RT_MULTI_DEVICE_SAFE_MODE` --> synced with the C++ flag through the functions.

**Review by gs-olive — COMMENTED** [review]
Add doc. to PR in runtime, which is linked to by the warning message.
"""
Detected run on multi-gpu system. Safety Doc w/threadpool mention
--> Perf implications of unsafe mode
"""

File: core/runtime/runtime.cpp [hunk 5]
```diff
@@ -31,13 +33,13 @@ c10::optional<RTDevice> get_most_compatible_device(const RTDevice& target_device
     if (device.device_name == target_device.device_name) {
       // First priority is selecting a candidate which agrees with the current device ID
       // If such a device is found, we can select it and break out of the loop
-      if (device.id == current_device.id && best_match.id != current_device.id) {
+      if (device.id == current_device.id) {
```
Comment (narendasan) [inline core/runtime/runtime.cpp:36]: Why are we changing these?

File: core/runtime/runtime.cpp [hunk 6]
```diff
@@ -103,6 +105,19 @@ RTDevice get_current_device() {
...
 }
 
+void multi_gpu_device_check(const RTDevice& most_compatible_device) {
+  // If multi-device safe mode is disabled and more than 1 device is registered on the machine, warn user
+  if (!(MULTI_DEVICE_SAFE_MODE) && get_available_device_list().get_devices().size() > 1) {
+    LOG_WARNING(
+        "Detected this engine is being instantitated in a multi-GPU system with "
+        << "multi-device safe mode disabled. For more on the implications of this "
+        << "as well as workarounds, see MULTI_DEVICE_SAFE_MODE.md "
+        << "(https://github.com/pytorch/TensorRT/blob/main/py/torch_tensorrt/dynamo/runtime/MULTI_DEVICE_SAFE_MODE.md). "
```
Comment (narendasan) [inline core/runtime/runtime.cpp:115]: Link to rendered docs on pytorch.org and make an rst

File: core/runtime/runtime.cpp (hunk 5 above)
Comment (gs-olive) [reply core/runtime/runtime.cpp:36]: During review, it was determined that these are not actually necessary conditions. Since the ID of each device is unique on a given machine, it should not be possible that `best_match.id == current_device.id`, so the check is not needed here.

File: docsrc/user_guide/multi_device_safe_mode.rst [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    with torch_tensorrt.runtime.set_multi_device_safe_mode(True):
+        ...
+
+TensorRT requires that each engine be associated with the CUDA context in the active thread from which it is invoked.
+Therefore, if the device were to change in the active thread, which may be the case when invoking
+engines on multiple GPUs from the same Python process, safe mode will cause Torch-TensorRT to display
+an alert and switch GPUs accordingly. If safe mode were not enabled, there could be a mismatch in the engine
+device and CUDA context device, which could lead the program to crash.
+
+One technique for managing multiple TRT engines on different GPUs while not sacrificing performance for
```
Comment (narendasan) [inline docsrc/user_guide/multi_device_safe_mode.rst:31]: Eventually we need a reference implementation of this

**Review by narendasan — APPROVED** [review]
LGTM, just some minor things that shouldnt require additional review

File: docsrc/user_guide/multi_device_safe_mode.rst (hunk 7 above)
Comment (gs-olive) [reply docsrc/user_guide/multi_device_safe_mode.rst:31]: Added task for this #2520
